src/infra/dex/balancer/dto.rs

```diff
@@ -1,58 +1,230 @@
+    sorGetSwapPaths(
+        callDataInput: $callDataInput,
+        chain: $chain,
+        queryBatchSwap: $queryBatchSwap,
+        swapAmount: $swapAmount,
+        swapType: $swapType,
+        tokenIn: $tokenIn,
+        tokenOut: $tokenOut,
+        useVaultVersion: $useVaultVersion
+    ) {
+        tokenAddresses
+        swaps {
+            poolId
+            assetInIndex
+            assetOutIndex
+            amount
+            userData
+        }
+        swapAmountRaw
+        returnAmountRaw
+        tokenIn
+        tokenOut
+    }
+}
+"#;
+
+#[derive(Serialize)]
 #[serde(rename_all = "camelCase")]
-pub enum OrderKind {
-    Sell,
-    Buy,
+pub struct Query<'a> {
+    query: &'a str,
+    variables: Variables,
+}
+
+impl Query<'_> {
+    pub fn from_domain(
+        order: &dex::Order,
+        tokens: &auction::Tokens,
+        slippage: &dex::Slippage,
+        chain_id: eth::ChainId,
+        contract_address: eth::ContractAddress,
+        query_batch_swap: bool,
+    ) -> Result<Self, Error> {
+        let token_decimals = match order.side {
+            order::Side::Buy => tokens
+                .decimals(&order.buy)
+                .ok_or(Error::MissingDecimals(order.buy)),
+            order::Side::Sell => tokens
+                .decimals(&order.sell)
+                .ok_or(Error::MissingDecimals(order.sell)),
+        }?;
+        let variables = Variables {
+            call_data_input: CallDataInput {
+                deadline: None,
+                receiver: contract_address.0,
+                sender: contract_address.0,
+                slippage_percentage: slippage.as_factor().clone(),
+            },
+            chain: Chain::from_domain(chain_id)?,
+            query_batch_swap,
+            swap_amount: HumanReadableAmount::from_decimal_units(
+                &order.amount.get(),
+                token_decimals,
+            ),
+            swap_type: SwapType::from_domain(order.side),
+            token_in: order.sell.0,
+            token_out: order.buy.0,
+            use_vault_version: Some(VaultVersion::V2.into()),
+        };
+        Ok(Self {
+            query: QUERY,
+            variables,
+        })
+    }
+}
+
+/// Refers to the SOR API V3's `AmountHumanReadable` type and represents a token
+/// amount without decimals.
+#[serde_as]
+#[derive(Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct HumanReadableAmount(BigDecimal);
+
+impl HumanReadableAmount {
+    /// Convert a `U256` amount to a human form.
+    pub fn from_decimal_units(units: &U256, decimals: u8) -> HumanReadableAmount {
+        let decimals: BigDecimal = BigInt::from(10).pow(decimals as u32).into();
+        Self(u256_to_big_decimal(units) / &decimals)
+    }
+
+    pub fn value(&self) -> &BigDecimal {
+        &self.0
+    }
+
+    /// Convert the human readable amount to a `U256` with 18 decimals.
+    pub fn to_decimal_units(&self, decimals: u8) -> Option<U256> {
```

Comment:
Seems a bit weird that this wrapper type need to know the decimals for all conversion operations. Maybe it makes more sense to store both the atoms and decimals?
Otherwise you could also create an instance of the struct with 6 decimals and read it back with 18 for example.

Reply:
Initially, a deserializer was required, but since this is not the case anymore, added the decimals field.